src/rest/DataUnionEndpoints.js

```diff
@@ -39,163 +267,382 @@ function throwIfBadAddress(address, variableDescription) {
+
+    log(`${collectedSignatureCount.toString()} out of ${requiredSignatureCount.toString()} collected`)
+    if (markedComplete) { log('All signatures collected') }
+    return markedComplete
+}
+
+/*
+function packSignatures(array) {
+    const msgLength = BigNumber.from(array.length).toHexString()
+    const [v, r, s] = array.reduce(([_v, _r, _s], e) => [_v.concat(e.v), _r.concat(e.r), _s.concat(e.s)], ['', '', ''])
+    return `${msgLength}${v}${r}${s}`
+}
+*/
+
+// move signatures from sidechain to mainnet
+async function transportSignatures(client, messageHash, options) {
+    const sidechainAmb = await getSidechainAmb(client, options)
+    const message = await sidechainAmb.message(messageHash)
+    const messageId = '0x' + message.substr(2, 64)
+    const sigCountStruct = await sidechainAmb.numMessagesSigned(messageHash)
+    const collectedSignatureCount = sigCountStruct.mask(255).toNumber()
+
+    log(`${collectedSignatureCount} signatures reported, getting them from the sidechain AMB...`)
+    const signatures = await Promise.all(Array(collectedSignatureCount).fill(0).map(async (_, i) => sidechainAmb.signature(messageHash, i)))
+
+    const [vArray, rArray, sArray] = [[], [], []]
+    signatures.forEach((signature, i) => {
+        log(`  Signature ${i}: ${signature} (len=${signature.length}=${signature.length / 2 - 1} bytes)`)
+        rArray.push(signature.substr(2, 64))
+        sArray.push(signature.substr(66, 64))
+        vArray.push(signature.substr(130, 2))
     })
-    const json = await response.json()
-    // server may return things like { code: "ConnectionPoolTimeoutException", message: "Timeout waiting for connection from pool" }
-    //   they must still be handled as errors
-    if (!response.ok && !json.error) {
-        json.error = `Server returned ${response.status} ${response.statusText}`
-    }
+    const packedSignatures = BigNumber.from(signatures.length).toHexString() + vArray.join('') + rArray.join('') + sArray.join('')
+    log(`All signatures packed into one: ${packedSignatures}`)
+
+    // Gas estimation also checks that the transaction would succeed, and provides a helpful error message in case it would fail
+    const mainnetAmb = await getMainnetAmb(client, options)
+    log(`Estimating gas using mainnet AMB @ ${mainnetAmb.address}, message=${message}`)
+    let gasLimit
+    try {
+        // magic number suggested by https://github.com/poanetwork/tokenbridge/blob/master/oracle/src/utils/constants.js
+        gasLimit = await mainnetAmb.estimateGas.executeSignatures(message, packedSignatures) + 200000
+        log(`Calculated gas limit: ${gasLimit.toString()}`)
+    } catch (e) {
+        // Failure modes from https://github.com/poanetwork/tokenbridge/blob/master/oracle/src/events/processAMBCollectedSignatures/estimateGas.js
+        log('Gas estimation failed: Check if the message was already processed')
+        const alreadyProcessed = await mainnetAmb.relayedMessages(messageId)
```

Comment:
What happens if signatures have already been transported? i.e. can anything bad happen if the message gets "already processed" after this check, but before the end of this method? i.e. async atomicity